File: stubs/csharp/src/BeastMode.RmDdd/Domain/AggregateBoundaries.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace BeastMode.RmDdd.Domain
{
    /// <summary>
    /// Defines aggregate boundaries and consistency rules.
    /// Specifies what entities belong to the aggregate and consistency constraints following .NET conventions.
    /// </summary>
    public class AggregateBoundaries
    {
        public string AggregateType { get; }
        public IReadOnlySet<string> EntityTypes { get; }
        public IReadOnlySet<string> Invariants { get; }
        public int MaxSize { get; }
        public bool EnforceConsistency { get; }
        
        private AggregateBoundaries(Builder builder)
        {
            AggregateType = builder.AggregateType ?? throw new ArgumentNullException(nameof(builder.AggregateType));
            EntityTypes = builder.EntityTypes.ToHashSet();
            Invariants = builder.Invariants.ToHashSet();
            MaxSize = builder.MaxSize;
            EnforceConsistency = builder.EnforceConsistency;
        }
        
        public bool ContainsEntityType(string entityType)
        {
            return EntityTypes.Contains(entityType);
        }
        
        public bool HasInvariant(string invariant)
        {
            return Invariants.Contains(invariant);
        }
        
        public static Builder Builder()
        {
            return new Builder();
        }
        
        public class Builder
        {
            internal string? AggregateType { get; private set; }
            internal HashSet<string> EntityTypes { get; } = new();
            internal HashSet<string> Invariants { get; } = new();
            internal int MaxSize { get; private set; } = 100; // Default max size
            internal bool EnforceConsistency { get; private set; } = true;
            
            public Builder WithAggregateType(string aggregateType)
            {
                AggregateType = aggregateType;
                return this;
            }
            
            public Builder WithEntityType(string entityType)
            {
                EntityTypes.Add(entityType);
                return this;
            }
            
            public Builder WithEntityTypes(IEnumerable<string> entityTypes)
            {
                foreach (var entityType in entityTypes)
                {
                    EntityTypes.Add(entityType);
                }
                return this;
            }
            
            public Builder WithInvariant(string invariant)
            {
                Invariants.Add(invariant);
                return this;
            }
            
            public Builder WithInvariants(IEnumerable<string> invariants)
            {
                foreach (var invariant in invariants)
                {
                    Invariants.Add(invariant);
                }
                return this;
            }
            
            public Builder WithMaxSize(int maxSize)
            {
                if (maxSize <= 0)
                    throw new ArgumentException("Max size must be positive", nameof(maxSize));
                
                MaxSize = maxSize;
                return this;
            }
            
            public Builder WithEnforceConsistency(bool enforceConsistency)
            {
                EnforceConsistency = enforceConsistency;
                return this;
            }
            
            public AggregateBoundaries Build()
            {
                return new AggregateBoundaries(this);
            }
        }
        
        public override bool Equals(object? obj)
        {
            if (obj is not AggregateBoundaries other) return false;
            if (ReferenceEquals(this, other)) return true;
            
            return AggregateType == other.AggregateType &&
                   EntityTypes.SetEquals(other.EntityTypes) &&
                   Invariants.SetEquals(other.Invariants) &&
                   MaxSize == other.MaxSize &&
                   EnforceConsistency == other.EnforceConsistency;
        }
        
        public override int GetHashCode()
        {
            return HashCode.Combine(AggregateType, EntityTypes, Invariants, MaxSize, EnforceConsistency);
        }
        
        public override string ToString()
        {
            return $"AggregateBoundaries {{ AggregateType = '{AggregateType}', " +
                   $"EntityTypes = [{string.Join(", ", EntityTypes)}], " +
                   $"Invariants = [{string.Join(", ", Invariants)}], " +
                   $"MaxSize = {MaxSize}, EnforceConsistency = {EnforceConsistency} }}";
        }
    }
}

File: stubs/java/src/main/java/com/beastmode/rmddd/domain/AggregateBoundaries.java
package com.beastmode.rmddd.domain;

import java.util.Objects;
import java.util.Set;
import java.util.HashSet;
import java.util.Collections;

/**
 * Defines aggregate boundaries and consistency rules.
 * Specifies what entities belong to the aggregate and consistency constraints.
 */
public class AggregateBoundaries {
    
    private final String aggregateType;
    private final Set<String> entityTypes;
    private final Set<String> invariants;
    private final int maxSize;
    private final boolean enforceConsistency;
    
    private AggregateBoundaries(Builder builder) {
        this.aggregateType = Objects.requireNonNull(builder.aggregateType, "Aggregate type cannot be null");
        this.entityTypes = Collections.unmodifiableSet(new HashSet<>(builder.entityTypes));
        this.invariants = Collections.unmodifiableSet(new HashSet<>(builder.invariants));
        this.maxSize = builder.maxSize;
        this.enforceConsistency = builder.enforceConsistency;
    }
    
    public String getAggregateType() {
        return aggregateType;
    }
    
    public Set<String> getEntityTypes() {
        return entityTypes;
    }
    
    public Set<String> getInvariants() {
        return invariants;
    }
    
    public int getMaxSize() {
        return maxSize;
    }
    
    public boolean isEnforceConsistency() {
        return enforceConsistency;
    }
    
    public boolean containsEntityType(String entityType) {
        return entityTypes.contains(entityType);
    }
    
    public boolean hasInvariant(String invariant) {
        return invariants.contains(invariant);
    }
    
    public static Builder builder() {
        return new Builder();
    }
    
    public static class Builder {
        private String aggregateType;
        private Set<String> entityTypes = new HashSet<>();
        private Set<String> invariants = new HashSet<>();
        private int maxSize = 100; // Default max size
        private boolean enforceConsistency = true;
        
        public Builder aggregateType(String aggregateType) {
            this.aggregateType = aggregateType;
            return this;
        }
        
        public Builder entityType(String entityType) {
            this.entityTypes.add(entityType);
            return this;
        }
        
        public Builder entityTypes(Set<String> entityTypes) {
            this.entityTypes.addAll(entityTypes);
            return this;
        }
        
        public Builder invariant(String invariant) {
            this.invariants.add(invariant);
            return this;
        }
        
        public Builder invariants(Set<String> invariants) {
            this.invariants.addAll(invariants);
            return this;
        }
        
        public Builder maxSize(int maxSize) {
            if (maxSize <= 0) {
                throw new IllegalArgumentException("Max size must be positive");
            }
            this.maxSize = maxSize;
            return this;
        }
        
        public Builder enforceConsistency(boolean enforceConsistency) {
            this.enforceConsistency = enforceConsistency;
            return this;
        }
        
        public AggregateBoundaries build() {
            return new AggregateBoundaries(this);
        }
    }
    
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        AggregateBoundaries that = (AggregateBoundaries) o;
        return maxSize == that.maxSize &&
               enforceConsistency == that.enforceConsistency &&
               Objects.equals(aggregateType, that.aggregateType) &&
               Objects.equals(entityTypes, that.entityTypes) &&
               Objects.equals(invariants, that.invariants);
    }
    
    @Override
    public int hashCode() {
        return Objects.hash(aggregateType, entityTypes, invariants, maxSize, enforceConsistency);
    }
    
    @Override
    public String toString() {
        return "AggregateBoundaries{" +
               "aggregateType='" + aggregateType + '\'' +
               ", entityTypes=" + entityTypes +
               ", invariants=" + invariants +
               ", maxSize=" + maxSize +
               ", enforceConsistency=" + enforceConsistency +
               '}';
    }
}